bgpd/bgp_mplsvpn.c

```diff
@@ -1336,6 +1336,230 @@ leak_update(struct bgp *to_bgp, struct bgp_dest *bn,
+		nh_resolved.prefixlen = IPV6_MAX_BITLEN;
+		IPV6_ADDR_COPY(&nh_resolved.u.prefix6,
+			       &bnc->nexthop->gate.ipv6);
+		nh_pfx = &nh_resolved;
+		break;
+	case NEXTHOP_TYPE_IFINDEX:
+		/* the nexthop is in bnc->prefix */
+		nh_pfx = &bnc->prefix;
+		break;
+	case NEXTHOP_TYPE_BLACKHOLE:
+		/* blachole routes do not support
+		 * per-nexthop label mode;
+		 * fallback to the per-vrf label.
+		 */
+		bgp_mplsvpn_path_nh_label_unlink(pi);
+		return from_bgp->vpn_policy[afi].tovpn_label;
+	}
+
+	/* one nexthop is connected
+	 * allocate a nexthop label cache entry
+	 */
+	tree = &from_bgp->mpls_labels_per_nexthop[family2afi(nh_pfx->family)];
+	blnc = bgp_label_per_nexthop_find(tree, nh_pfx);
+	if (!blnc) {
+		blnc = bgp_label_per_nexthop_new(tree, nh_pfx);
+		blnc->to_bgp = to_bgp;
+		/* request a label to zebra for this nexthop
+		 * the response from zebra will trigger the callback
+		 */
+		bgp_lp_get(LP_TYPE_NEXTHOP, blnc,
+			   bgp_mplsvpn_get_label_per_nexthop_cb);
+	}
+
+	if (pi->label_nexthop_cache == blnc)
+		/* no change */
+		return blnc->label;
+
+	/* Unlink from existing nexthop cache, if any.
+	 * This will also free the nexthop label cache entry
+	 */
+	bgp_mplsvpn_path_nh_label_unlink(pi);
+	if (blnc) {
+		/* updates NHT pi list reference */
+		LIST_INSERT_HEAD(&(blnc->paths), pi, label_nh_thread);
+		pi->label_nexthop_cache = blnc;
+		pi->label_nexthop_cache->path_count++;
+	}
+
+	return blnc->label;
+}
+
+static mpls_label_t vpn_leak_from_vrf_get_label_for_path_vrf(
+	afi_t afi, safi_t safi, struct bgp_path_info *pi, struct bgp *from_bgp,
+	struct bgp *to_bgp)
+{
+	mpls_label_t label_val = from_bgp->vpn_policy[afi].tovpn_label;
+
+	if (!CHECK_FLAG(from_bgp->vpn_policy[afi].flags,
+			BGP_VPN_POLICY_TOVPN_LABEL_PER_NEXTHOP))
+		/* per label nexthop mode not configured.
+		 * The per VRF mode is selected.
+		 * return the per VRF label
+		 */
+		return label_val;
+
+	/* per-label nexthop mode is configured */
+	if (afi == AFI_IP) {
+		if (CHECK_FLAG(pi->attr->flag,
+			       ATTR_FLAG_BIT(BGP_ATTR_NEXT_HOP)) &&
+		    pi->attr->nexthop.s_addr != INADDR_ANY &&
+		    ipv4_unicast_valid(&pi->attr->nexthop))
+			label_val = bgp_mplsvpn_get_path_info_label(
+				pi, to_bgp, from_bgp, afi, safi);
+		else if ((pi->attr->mp_nexthop_len == BGP_ATTR_NHLEN_IPV4) &&
+			 pi->attr->mp_nexthop_global_in.s_addr != INADDR_ANY &&
+			 ipv4_unicast_valid(&pi->attr->mp_nexthop_global_in))
+			label_val = bgp_mplsvpn_get_path_info_label(
+				pi, to_bgp, from_bgp, afi, safi);
+	} else if (afi == AFI_IP6 &&
+		   (pi->attr->mp_nexthop_len == BGP_ATTR_NHLEN_IPV6_GLOBAL ||
+		    pi->attr->mp_nexthop_len ==
+			    BGP_ATTR_NHLEN_IPV6_GLOBAL_AND_LL) &&
+		   !IN6_IS_ADDR_UNSPECIFIED(&pi->attr->mp_nexthop_global) &&
+		   !IN6_IS_ADDR_LOOPBACK(&pi->attr->mp_nexthop_global) &&
+		   !IN6_IS_ADDR_MULTICAST(&pi->attr->mp_nexthop_global))
+		label_val = bgp_mplsvpn_get_path_info_label(
+			pi, to_bgp, from_bgp, afi, safi);
+
+	return label_val;
```

Comment:
describe why you are falling back to per vrf label

Reply:
done:
```
/* nexthop from pi can not be used with per-nexthop label:
* - "network" configured prefixes in "no bgp network import-check" mode
* - redistributed routes
* - martian nexthops
* Fallback to the per VRF label.
*/
```